### ISSUE TITLE
Use model to calculate expected improvement for a list of candidates

### ISSUE THREAD

**Issue by sgbaird** [facebook/Ax#771]
The idea would be to constrain Ax to only "suggest next experiments" from a predefined list of candidates using a `model` output (e.g. from `optimize()`) without that constraint? I would expect this to take the form of calculating the expected improvement for each of the candidates and returning the candidate with the maximum expected improvement.

**Comment by Balandat** [comment]
Seems like a good idea. The `Model` API already has an `evaluate_acquisition_function` method: https://github.com/facebook/Ax/blob/main/ax/models/torch_base.py#L220. However, that acts on a tensor input, i.e. on the transformed candidates. But it should be pretty straightforward to surface this directly in the modelbridge and allow evaluation of the acquisition function of standard parameterizations. Then picking the best out of those is simple.

**Comment by lena-kashtelyan** [comment]
Actually, there is a way to `evaluate_acquisition_function` directly on a point in experiment search space via [`ModelBridge.evaluate_acquisition_function`](https://github.com/facebook/Ax/blob/main/ax/modelbridge/base.py#L825), which wraps `BoTorchModel.evaluate_acquisition_function`! The model will need to be `Models.BOTORCH_MODULAR` to make use of this, but it should work. For `AxClient`, you'd do this like so, @sgbaird:

```
from ax.core.arm import Arm
from ax.core.observation import ObservationFeatures
from ax.modelbridge.modelbridge_utils import extract_search_space_digest

model_bridge = ax_client.generation_strategy.model
search_space_digest = extract_search_space_digest(
      search_space=ax_client.experiment.search_space, 
      param_names=model_bridge.parameters
  )
objective_weights = extract_objective_weights(
      objective=ax_client.experiment.optimization_config.objective, 
      outcomes=model_bridge.outcomes
  )

# `acqf_values` is a list of floats (since we can evaluate acqf for multiple points at once); 
# ordering corresponds to order of points in `observation_features` input
acqf_values = model_bridge.evaluate_acquisition_function(
     # Each `ObservationFeatures` below represents one point in experiment (untransformed search space):
    observation_features=[ObservationFeatures.from_arm(Arm(parameters={"x": ..., ...})), ...],
    search_space_digest=search_space_digest,
    objective_weights=objective_weights,
)
```

I think it's actually unnecessary for `evaluate_acquisition_function` to require `search_space_digest` and `objective_weights`, we'll fix that (so you don't need to manually construct them and so doing this is a lot simpler), but in the meantime this code block should work!

**Comment by Balandat** [comment]
Ah, must have somehow missed that, good catch.

**Comment by lena-kashtelyan** [comment]
@sgbaird, let us know how this works out for you!

**Comment by sgbaird** [comment]
@lena-kashtelyan and @Balandat, thank you! I will give it a try and let you know.

**Comment by sgbaird** [comment]
I've been getting started on this and I think it will work out. One thing I'm trying to figure out though is what kind of reasonable upper limit I can get with number of candidates.

**Comment by lena-kashtelyan** [comment]
> I've been getting started on this and I think it will work out

Sweet!

> One thing I'm trying to figure out though is what kind of reasonable upper limit I can get with number of candidates.

I imagine the upper bound will be performance, but cc @dme65 and @Balandat for other thoughts!

**Comment by sgbaird** [comment]
@lena-kashtelyan, I'm having some trouble. It seems that `model_bridge == None` unless I use something like `ax_client.get_next_trial()`, however after doing that, then I get `NotImplementedError` for `evaluate_acquisition_function`.

Attached is a long reproducer ([service_predefined_debug.txt](https://github.com/facebook/Ax/files/7859551/service_predefined_debug.txt)) and here is the more relevant snippet from the reproducer:
```python
ax_client = AxClient(generation_strategy=gs)
ax_client.create_experiment(
    name=exp_name,
    parameters=orig_parameters,
    parameter_constraints=[orig_comp_constraint],
    objective_name=target_name,
    minimize=True,
    immutable_search_space_and_opt_config=False,
)

ct = 0
for i in range(n_train):
    ax_client.attach_trial(X_train.iloc[i, :-1].to_dict())
    ax_client.complete_trial(trial_index=ct, raw_data=(y_train[i], 5.0))
    ct = ct + 1

y_preds = []
y_stds = []
next_experiments = []
for i in range(batch_size):
    # ax_client.get_next_trial() # by uncommenting this, model_bridge is assigned, but evaluate_acquisition_function is NotImplemented
    model_bridge = ax_client.generation_strategy.model
    transformed_gen_args = model_bridge._get_transformed_gen_args(
        search_space=ax_client.experiment.search_space,
    )
    search_space_digest = extract_search_space_digest(
        search_space=transformed_gen_args.search_space,
        param_names=model_bridge.parameters,
    )
    objective_weights = extract_objective_weights(
        objective=ax_client.experiment.optimization_config.objective,
        outcomes=model_bridge.outcomes,
    )

    observation_features = [
        ObservationFeatures(dict(zip(unique_components[:-1], x_pred)))
        for x_pred in X_pred.iloc[:, :-1].values
    ]

    # `acqf_values` is a list of floats (since we can evaluate acqf for multiple points at once);
    # ordering corresponds to order of points in `observation_features` input
    acqf_values = model_bridge.evaluate_acquisition_function(
        # Each `ObservationFeatures` below represents one point in experiment (untransformed) search space:
        observation_features=observation_features,
        search_space_digest=search_space_digest,
        objective_weights=objective_weights,
    )

    max_value = max(acqf_values)
    max_index = acqf_values.index(max_value)
    next_experiment = X_pred.iloc[max_index]

    model = ax_client.generation_strategy.model
    y_pred, y_var = model.predict([ObservationFeatures(next_experiment)])
    y_pred = y_pred[target_name][0]
    y_var = y_var[target_name][target_name][0]
    y_std = np.sqrt(y_var)

    y_preds.append(y_pred)
    y_stds.append(y_std)
    next_experiments.append(next_experiment)
    ct = ct + 1
```

**Comment by lena-kashtelyan** [comment]
Ah yes, sorry I failed to mention an important piece! You'll need to specify the generation strategy like this one: https://ax.dev/tutorials/modular_botax.html#5.-Utilizing-BoTorchModel-in-generation-strategies, since `evaluate_acquisition_function` is only available on `Models.BOTORCH_MODULAR`. It will also only be possible after the first few Sobol trials. Let me know if this helps!

**Comment by Balandat** [comment]
> One thing I'm trying to figure out though is what kind of reasonable upper limit I can get with number of candidates.

So IIUC here you're just looking to compute the acquisition value for each candidate individually, rather than compute the joint acquisition value of a set of points. In that case there shouldn't be any major scalability issues. At least down on the botorch end we routinely compute acquisition values for thousands of candidates in parallel. The good thing here is that you only need to evaluate the acquisition function once (rather than many times as you would need to for optimizing an acquisition function over some continuous space).

**Comment by sgbaird** [comment]
That seemed to do the trick! In addition to using `Models.BOTORCH_MODULAR` and to get around needing to generate SOBOL points (i.e. use my own data to start out with), I made a dummy trial via:
```python
# ensure ax_client.generation_strategy.model gets assigned (then discard dummy trial)
_, trial_index = ax_client.get_next_trial()
ct = ct + 1
trial = ax_client.experiment.trials[trial_index]
trial.mark_completed()
```
I think this counts as marking the trial as failed, but not abandoned, correct? https://github.com/facebook/Ax/issues/372#issuecomment-682660998

Here is the updated script: [service_predefined_debug.txt](https://github.com/facebook/Ax/files/7871729/service_predefined_debug.txt).

**Comment by sgbaird** [comment]
I get an interesting error related to the max number of candidates:
```python
SobolQMCSampler only supports dimensions `q * o <= 21201`. Requested: 32963
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\botorch\sampling\samplers.py", line 297, in _construct_base_samples
    raise UnsupportedError(
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\botorch\sampling\samplers.py", line 92, in forward
    self._construct_base_samples(posterior=posterior, shape=base_sample_shape)
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\botorch\acquisition\monte_carlo.py", line 245, in forward
    samples = self.sampler(posterior)
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\botorch\utils\transforms.py", line 220, in decorated
    output = method(acqf, X, *args, **kwargs)
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\botorch\utils\transforms.py", line 261, in decorated
    return method(cls, X, **kwargs)
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\ax\models\torch\botorch_modular\acquisition.py", line 316, in evaluate
    return self.acqf(X=X)
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\ax\models\torch\botorch_modular\model.py", line 341, in evaluate_acquisition_function
    return acqf.evaluate(X=X)
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\ax\modelbridge\torch.py", line 134, in _model_evaluate_acquisition_function
    evals = not_none(self.model).evaluate_acquisition_function(
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\ax\modelbridge\array.py", line 494, in _evaluate_acquisition_function
    return self._model_evaluate_acquisition_function(
  File "C:\Users\sterg\miniconda3\envs\Ultradent-BO\Lib\site-packages\ax\modelbridge\base.py", line 866, in evaluate_acquisition_function
    return self._evaluate_acquisition_function(
  File "C:\Users\sterg\Documents\GitHub\sparks-issa\Ultradent-BO\service_predefined.py", line 209, in <module>
    acqf_values = model_bridge.evaluate_acquisition_function(
```
I'm curious if `acqf_values` depend on each other or are stochastic, or if the SOBOL generation underneath the hood is just extra computation happening. Hard-coding a higher max caused a large slowdown on my computer - ended up exiting the program (I'm assuming this has to do with large matrices, e.g. 50000 x 50000 takes a lot of memory).

**Comment by lena-kashtelyan** [comment]
> In addition to using Models.BOTORCH_MODULAR and to get around needing to generate SOBOL points (i.e. use my own data to start out with)

Oh, in this case you can actually just omit the Sobol step from your generation strategy! 

> I think this counts as marking the trial as failed, but not abandoned, correct? #372 (comment)

I think in your example, you are just marking trial as completed but without data. So it's not failed, but just kind of ignored. To actually mark a trial as failed, you can use `ax_client.mark_trial_failed` , but if you omit the Sobol step altogether from your generation strategy, you won't need to do that : )

**Comment by lena-kashtelyan** [comment]
Re: the error you are getting –– that's interesting, let us look into it! cc @Balandat and @dme65 on this one

> Also, for some reason, I'm getting a single value (wrapped in a list) as the output for acqf_values instead of a list of the same size as the number of candidates.

Strange, I'll look into that!

**Comment by sgbaird** [comment]
I have a workaround for the single value output bug:
```python
    acqf_values = [
        model_bridge.evaluate_acquisition_function(
            # Each `ObservationFeatures` below represents one point in experiment (untransformed) search space:
            observation_features=[observation_feature],
            search_space_digest=search_space_digest,
            objective_weights=objective_weights,
        )[0]
        for observation_feature in observation_features
    ]
```
which incidentally avoids the issue of having too many points; however, it seems pretty slow.

**Comment by lena-kashtelyan** [comment]
Let me actually open a separate issue just for that bug!

**Comment by sgbaird** [comment]
@lena-kashtelyan also, is it possible to use `BOTORCH_MODULAR` with `FULLYBAYESIAN` (i.e. [SAASBO](https://ax.dev/tutorials/saasbo.html)) instead of `SingleTaskGP`? I had trouble finding SAASBO within BoTorch as a model similar to `SingleTaskGP`.

**Comment by dme65** [comment]
> @lena-kashtelyan also, is it possible to use `BOTORCH_MODULAR` with `FULLYBAYESIAN` (i.e. [SAASBO](https://ax.dev/tutorials/saasbo.html)) instead of `SingleTaskGP`? I had trouble finding SAASBO within BoTorch as a model similar to `SingleTaskGP`.

SAASBO doesn't exist as a pure BoTorch model, but rather as an Ax BoTorch model. The relevant code is here: https://github.com/facebook/Ax/blob/main/ax/models/torch/fully_bayesian.py. 

Moving SAASBO to BoTorch and making it available in modular BoTorch is in the works, but currently isn't available.

**Comment by lena-kashtelyan** [comment]
@sgbaird, are you blocked on the ability to evaluate acquisition function while also using SAASBO?

**Comment by sgbaird** [comment]
@lena-kashtelyan yes. SAASBO seems a lot better suited to what I'm working on (small training dataset size, very expensive trials, limited to a couple hundred additional trials, and high dimensionality usually reduced to fewer out of necessity - 20+ dimensions). So far evaluating the acquisition function for predefined candidates is the only way I've been able to reliably meet a `max_components` constraint. I also understand that there are a lot of competing priorities in Ax development. Happy to help where I can.

**Comment by sgbaird** [comment]
Something else important I discovered about the example - I need to use `deepcopy` on ax_client otherwise the parameter bounds seem to be overwritten with the range [0.0, 1.0]:
```python
from copy import deepcopy
```
```python
    ax_client_copy = deepcopy(ax_client)
    model_bridge = ax_client_copy.generation_strategy.model
    transformed_gen_args = model_bridge._get_transformed_gen_args(
        search_space=ax_client_copy.experiment.search_space,
    )
    search_space_digest = extract_search_space_digest(
        search_space=transformed_gen_args.search_space,
        param_names=model_bridge.parameters,
    )
    objective_weights = extract_objective_weights(
        objective=ax_client_copy.experiment.optimization_config.objective,
        outcomes=model_bridge.outcomes,
    )
```

**Comment by mcd01** [comment]
I am having a follow-up question with regards to the question discussed in this thread. So, in case one wants to add initial data to a freshly created model and then calculate EI for a list of candidates (i.e. the use case of @sgbaird ), there is no need for actually getting a (potentially heavy to compute) trial suggestion from the search space. Yet, this needs to be done here in order to generate the model internally (see https://github.com/facebook/Ax/issues/771#issuecomment-1013287790), and I am also experiencing this in my own setup. So the question would be, given that one uses the Service Api, is there a simple way of triggering model fitting / updating (after having manually attached some trials) without actually requesting a new trial suggestion (so that one can directly proceed with the EI calculcation for a list of candidates)?

Just asking out of curiosity, if you have an idea, let me know, I am grateful for any input :)

**Comment by Balandat** [comment]
> given that one uses the Service Api, is there a simple way of triggering model fitting / updating (after having manually attached some trials) without actually requesting a new trial suggestion

I don't believe the Service API currently support this. We do have a `get_model_predictions` interface, but interesting that function errors out if called before candidates have been generated: https://github.com/facebook/Ax/blob/main/ax/service/ax_client.py#L931-L950

I guess it wouldn't be hard to just fit the model instead of erroring out, that way if one calls `get_model_predictions` on one of the existing arms the model would be fitted on the backend (necessary and less costly than generating a candidate) and one could compute the acquisition function after.

It seems that currently `get_model_predictions` also only works on existing arms; we should probably add and API for predicting "out-of-sample" configurations. cc @lena-kashtelyan

**Comment by lena-kashtelyan** [comment]
Linking:
- separate issue for SAASBO in modular BotAx (MBM): https://github.com/facebook/Ax/issues/795,
- issue for exposing `AxClient.fit_model` and extending `AxClient.get_model_predictions` to support out-of-sample points using it: https://github.com/facebook/Ax/issues/819.

I believe with this in mind, the only outstanding issue was making `evaluate_acquisition_function` on `ModelBridge` more convenient; that is now on latest and will be included in the imminent stable version release, so marking this as "fixready".

**Comment by sgbaird** [comment]
@lena-kashtelyan thank you!
